File: frontend/src/types/renewal.ts
import type { AssetRecord } from './asset';

export interface RenewalItemRecord {
  id: number;
  renewal_id: number;
  asset_id: number;
  new_supplier_id?: number | null;
  new_expiry_date?: string | null;
  new_cost?: number;
}

export interface RenewalRecord {
  id: number;
  renewal_no: string;
  status: 'pending' | 'completed';
  created_at?: string;
  updated_at?: string;
  items?: RenewalItemRecord[];
}

export interface CreateRenewalPayload {
  renewal_no: string;
  asset_ids: number[];
}

export interface CompleteRenewalPayloadItem {
  asset_id: number;
  new_supplier_id?: number;
  new_cost: number;
  new_date: string;
}

export interface CompleteRenewalPayload {
  renewal_id: number;
  items: CompleteRenewalPayloadItem[];
}

export interface RenewalDraftItem {
  asset: AssetRecord;
  asset_id: number;
  new_supplier_id?: number;
  new_cost?: number;
  new_expiry_date?: string;
  new_next_billing_date?: string;
}

export interface RenewalDraft {
  renewal_id: number;
  renewal_no: string;
  items: RenewalDraftItem[];
}

File: backend/models/renewal.go
package models

import "time"

type Renewal struct {
	ID        uint          `json:"id" gorm:"primaryKey"`
	RenewalNo string        `json:"renewal_no" gorm:"size:100;uniqueIndex;not null"`
	Status    string        `json:"status" gorm:"size:20;not null;default:pending"`
	CreatedAt time.Time     `json:"created_at"`
	UpdatedAt time.Time     `json:"updated_at"`
	Items     []RenewalItem `json:"items,omitempty" gorm:"constraint:OnDelete:CASCADE"`
}

type RenewalItem struct {
	ID            uint       `json:"id" gorm:"primaryKey"`
	RenewalID     uint       `json:"renewal_id" gorm:"index;not null"`
	AssetID       uint       `json:"asset_id" gorm:"index;not null"`
	NewSupplierID *uint      `json:"new_supplier_id"`
	NewExpiryDate *time.Time `json:"new_expiry_date" gorm:"type:date"`
	NewCost       float64    `json:"new_cost"`

	Asset       *Asset    `json:"asset,omitempty"`
	NewSupplier *Supplier `json:"new_supplier,omitempty"`
}
